fix(parser): store lone surrogates as escape sequence

overlookmotel opened the pull request:
Closes #3526.

Fully parse strings containing lone surrogates and encode the string in `value`.

Encoding schema is to encode a lone surrogate as the lossy replacement character, followed by the code point in hex. i.e. `"\uD800"` in source code would be encoded as `\u{FFFD}d800` in `value`. The lossy replacement character itself is encoded as `\u{FFFD}fffd`.

There's nothing special about the lossy replacement character. Just had to choose *some* valid Unicode character to be the escape marker, and that seemed like a reasonable choice of a character which is likely to be rare in real-world code.

All the ESTree-Test262 test cases related to lone surrogates now pass.

WIP. A bit of tidying up to do yet.

overlookmotel commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/10041?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#10041** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/10041?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/10041?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#10040** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/10040?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#10039** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/10039?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#10036** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/10036?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=10041).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

Boshen commented:
cc @andreubotella for a quick review.

andreubotella commented:
> cc @andreubotella for a quick review.

IIUC, if `lone_surrogates` is false, then any replacement characters in the value are true replacement characters; and if true, then the value string would need decoding, right? I think we can work with that in Nova. (cc @aapoalas)

That said, I worry that other users of oxc_ast might not be aware of this change, and might use `value` without checking for `lone_surrogate`. You could say that no `AsRef<str>` would accurately represent the JS string, but I'd say that decoding any lone surrogate lossily to the replacement character (as `String::from_utf16_lossy` does) would be a lot more useful for most use cases. If the extra memory use is not an issue, I'd suggest having `value` be this lossy decode, and having `lone_surrogate_string` be an `Option<Atom<'a>>` that is only `Some` if there are lone surrogates.

overlookmotel commented:
@andreubotella Thanks for your review.

Yes, this isn't entirely ideal for the 2 reasons you've pointed out:

1. The extra `lone_surrogates` flag adds 8 bytes to `StringLiteral`.
2. It's confusing that `str_lit.value.as_str()` sometimes returns an inaccurate string - you have to check the `lone_surrogates` flag to see if it needs to be decoded.

I considered the main priority to encode the string correctly *somehow* and to be able to pass the Test262 tests.

But I see this as a stop-gap. In the end, I think we need our own string types, to enable some other optimizations:

* String length is always `<= u32::MAX`.
* Separate `Ident` type for identifiers, which stores a 32-bit hash inline, to make using them as hash table keys faster.
* Faster string comparison and hashing.

https://github.com/oxc-project/backlog/issues/46#issuecomment-2603013242
https://github.com/oxc-project/backlog/issues/46#issuecomment-2603153807

So, when we do that, I'd imagine changing the implementation to more like:

```rs
struct StringLiteralValue {
    ptr: NonNull<u8>,
    len: u32,
    lone_surrogates: bool,
    _padding: [u8; 3],
}
```

`StringLiteralValue` would not dereference to `&str`, and when you print it, it would automatically decode the escapes. I think this would address both shortcomings.

But in meantime, I propose that we go with the approach in this PR, as it's at least is correct, even if it's messy. Does that sound OK to you?

overlookmotel commented:
Something is wrong. This PR should make no difference to result of `cargo minsize`, because codegen is only using the value of `raw` when `lone_surrogates` flag is set (prior to #10044). I'll investigate.

overlookmotel commented:
> Something is wrong. This PR should make no difference to result of `cargo minsize`, because codegen is only using the value of `raw` when `lone_surrogates` flag is set (prior to #10044). I'll investigate.

Nope. Snapshot was just out of date.

overlookmotel committed:
fix(parser): store lone surrogates as escape sequence (#10041)

Closes #3526.

Fully parse strings containing lone surrogates and encode the string in `value`.

Encoding schema is to encode a lone surrogate as the lossy replacement character, followed by the code point in hex. i.e. `"\uD800"` in source code would be encoded as `\u{FFFD}d800` in `value`. The lossy replacement character itself is encoded as `\u{FFFD}fffd`.

There's nothing special about the lossy replacement character. Just had to choose *some* valid Unicode character to be the escape marker, and that seemed like a reasonable choice of a character which is likely to be rare in real-world code.

All the ESTree-Test262 test cases related to lone surrogates now pass.

WIP. A bit of tidying up to do yet.

overlookmotel commented:
@overlookmotel Note to self: Add some tests. Especially for strings containing both lone surrogates and lossy replacement characters.

overlookmotel commented:
Tests added in #10175.
